cloudinit/sources/DataSourceAzure.py

```diff
@@ -1387,76 +1387,95 @@ def load_azure_ds_dir(source_dir):
+
+
+@azure_ds_telemetry_reporter
+def _generate_network_config_from_imds_metadata(imds_metadata) -> dict:
+    """Convert imds_metadata dictionary to network v2 configuration.
+    Parses network configuration from imds metadata.
+
+    @param: imds_metadata: Dict of content read from IMDS network service.
+    @return: Dictionary containing network version 2 standard configuration.
+    """
+    netconfig = {'version': 2, 'ethernets': {}}
+    network_metadata = imds_metadata['network']
+    for idx, intf in enumerate(network_metadata['interface']):
+        # First IPv4 and/or IPv6 address will be obtained via DHCP.
+        # Any additional IPs of each type will be set as static
+        # addresses.
+        nicname = 'eth{idx}'.format(idx=idx)
+        dhcp_override = {'route-metric': (idx + 1) * 100}
+        dev_config = {'dhcp4': True, 'dhcp4-overrides': dhcp_override,
+                      'dhcp6': False}
+        for addr_type in ('ipv4', 'ipv6'):
+            addresses = intf.get(addr_type, {}).get('ipAddress', [])
+            if addr_type == 'ipv4':
+                default_prefix = '24'
+            else:
+                default_prefix = '128'
+                if addresses:
+                    dev_config['dhcp6'] = True
+                    # non-primary interfaces should have a higher
+                    # route-metric (cost) so default routes prefer
+                    # primary nic due to lower route-metric value
+                    dev_config['dhcp6-overrides'] = dhcp_override
+            for addr in addresses[1:]:
+                # Append static address config for ip > 1
+                netPrefix = intf[addr_type]['subnet'][0].get(
+                    'prefix', default_prefix)
+                privateIp = addr['privateIpAddress']
+                if not dev_config.get('addresses'):
+                    dev_config['addresses'] = []
+                dev_config['addresses'].append(
+                    '{ip}/{prefix}'.format(
+                        ip=privateIp, prefix=netPrefix))
+        if dev_config:
+            mac = ':'.join(re.findall(r'..', intf['macAddress']))
+            dev_config.update({
+                'match': {'macaddress': mac.lower()},
+                'set-name': nicname
+            })
+            # With netvsc, we can get two interfaces that
+            # share the same MAC, so we need to make sure
+            # our match condition also contains the driver
+            driver = device_driver(nicname)
+            if driver and driver == 'hv_netvsc':
+                dev_config['match']['driver'] = driver
+            netconfig['ethernets'][nicname] = dev_config
+    return netconfig
+
+
+@azure_ds_telemetry_reporter
+def _generate_network_config_from_fallback_config() -> dict:
+    """Generate fallback network config excluding mlx4_core & mlx5_core devices.
+
+    @return: Dictionary containing network version 2 standard configuration.
+    """
+    blacklist = ['mlx4_core', 'mlx5_core']
```

Comment:
@johnsonshi Thank you very much for this update. I was just talking today about how we needed this blacklist_driver update in tip of master due to my SRU verification of #539. Can you add a comment a suggested known instance type that should use mlx5_core driver so it'll ease testing and verification for our next SRU? I thought it was any v3 type advanced networking, but I wasn't sure.

Reply:
I've just checked with the internal teams that handle Azure VM Instance SKUs. They've said that the [Dv4 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/dv4-dsv4-series) and the [Ev4 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/ev4-esv4-series) always have `mlx5`. Older Dv-? and Ev-? series VMs (such as [Dv2 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/dv2-dsv2-series), [Dv3 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/dv3-dsv3-series), and [Ev3 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/ev3-esv3-series)) may have either `mlx5` or `mlx4` hardware (with the older series ones having a higher chance of older `mlx4` hardware). I'll update the comment above with suggested instance types.

Reply:
Comment added. Thanks @blackboxsw!